datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java

```diff
@@ -141,6 +141,116 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
     return newAssignment;
   }
 
+  /**
+   * Move a partition for a datastream group according to the targetAssignment. As we are only allowed to mutate the
+   * task once. It follow the steps
+   * Step 1) get the partitions that to be moved, and get their source task
+   * Step 2) If the instance is the instance we want to move, we find out the task that we want to assign the task
+   * Step 3) We scan the old task, compute new task if the old task belongs to these source tasks or if it is the
+   * target task we want to move to
+   *
+   * @param currentAssignment the old assignment
+   * @param targetAssignment the target assignment retrieved from Zookeeper
+   * @param partitionsMetadata the subscribed partitions metadata received from connector
+   * @return new assignment
+   */
+  public Map<String, Set<DatastreamTask>> movePartitions(Map<String, Set<DatastreamTask>> currentAssignment,
+      Map<String, Set<String>> targetAssignment, DatastreamGroupPartitionsMetadata partitionsMetadata) {
+
+    LOG.info("Move partition, task: {}, target assignment: {}, all partitions: {}", currentAssignment,
+        targetAssignment, partitionsMetadata.getPartitions());
+
+    String dgName = partitionsMetadata.getDatastreamGroup().getName();
+
+    Set<String> allToReassignPartitions = new HashSet<>();
+    targetAssignment.values().stream().forEach(allToReassignPartitions::addAll);
+    allToReassignPartitions.retainAll(partitionsMetadata.getPartitions());
+
+    //construct a map to store the tasks and if it contain the partitions that need to be released
+    //map: <taskName, partitions that need to be released>
+    Map<String, Set<String>> toReleasePartitionsTaskMap = new HashMap<>();
+    Map<String, String> partitionToSourceTaskMap = new HashMap<>();
+
+    //We first confirmed that the partitions in the target assignment can be removed, and we found its source task
+    currentAssignment.keySet().stream().forEach(instance -> {
+      Set<DatastreamTask> tasks = currentAssignment.get(instance);
+      tasks.forEach(task -> {
+        if (dgName.equals(task.getTaskPrefix())) {
+          Set<String> toMovePartitions = new HashSet<>(task.getPartitionsV2());
+          toMovePartitions.retainAll(allToReassignPartitions);
+          toReleasePartitionsTaskMap.put(task.getDatastreamTaskName(), toMovePartitions);
+          toMovePartitions.forEach(p -> partitionToSourceTaskMap.put(p, task.getDatastreamTaskName()));
+        }
+      });
+    });
+
+    Set<String> tasksToMutate = toReleasePartitionsTaskMap.keySet();
```

Comment:
toReleasePartitionsTaskMap that contains empty values doesn't need mutation right?